core/src/main/java/org/apache/iceberg/DistributedDataBatchScan.java

```diff
@@ -0,0 +1,309 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.ManifestEvaluator;
+import org.apache.iceberg.expressions.Projections;
+import org.apache.iceberg.expressions.ResidualEvaluator;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.MoreExecutors;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.iceberg.util.TableScanUtil;
+
+abstract class DistributedDataBatchScan
+    extends SnapshotScan<BatchScan, ScanTask, ScanTaskGroup<ScanTask>> implements BatchScan {
+
+  private static final int NUM_LOCAL_CORES = Runtime.getRuntime().availableProcessors();
+  private static final long METADATA_LIMIT_PER_LOCAL_CORE = 128 * 1024 * 1024; // 128 MB
+  private static final int NUM_WORKER_THREADS = 2;
+
+  protected DistributedDataBatchScan(Table table, Schema schema, TableScanContext context) {
+    super(table, schema, context);
+  }
+
+  protected abstract int remoteParallelism();
+
+  protected abstract PlanningMode dataPlanningMode();
+
+  protected abstract List<DataFile> planDataRemotely(List<ManifestFile> manifests);
+
+  protected abstract PlanningMode deletePlanningMode();
+
+  protected abstract List<DeleteFile> planDeletesRemotely(List<ManifestFile> manifests);
+
+  protected abstract void releaseResources();
+
+  @Override
+  protected boolean useSnapshotSchema() {
+    return true;
+  }
+
+  @Override
+  protected CloseableIterable<ScanTask> doPlanFiles() {
+    Snapshot snapshot = snapshot();
+
+    List<ManifestFile> dataManifests = findMatchingDataManifests(snapshot);
+    List<ManifestFile> deleteManifests = findMatchingDeleteManifests(snapshot);
+
+    boolean planDataRemotely = shouldPlanRemotely(dataPlanningMode(), dataManifests);
+    boolean planDeletesRemotely = shouldPlanRemotely(deletePlanningMode(), deleteManifests);
+
+    if (planDataRemotely || planDeletesRemotely) {
+      ExecutorService workerPool = newWorkerPool();
```

Comment:
This is more of a monitor pool, right? It is used just to monitor the parallel reads?

Reply:
Yeah, the name is probably misleading. It is purely to submit data and deletes future concurrently.